### PR TITLE
fix for Decimal128 group by [issue-3378]

### PR DESCRIPTION
**Pull request by 4ertus2**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

### CODE REVIEW

File: dbms/src/Columns/ColumnDecimal.cpp [hunk 1]
```diff
@@ -117,6 +121,36 @@ MutableColumnPtr ColumnDecimal<T>::cloneResized(size_t size) const
...
+
+    const void * src = pos;
+    void * dst = &data.back();
+
+#if __SSE2__
+    /// prevent aligned SSE load/store.
+    if constexpr (sizeof(T) == 4)
+    {
+        __m128 value = _mm_load_ss(static_cast<const float *>(src));
+        _mm_store_ss(static_cast<float *>(dst), value);
```
Comment (alexey-milovidov) [inline dbms/src/Columns/ColumnDecimal.cpp]: Why so complicated code for 4 and 8 bytes copy?

File: dbms/src/Columns/ColumnDecimal.cpp [hunk 2]
```diff
@@ -117,6 +121,36 @@ MutableColumnPtr ColumnDecimal<T>::cloneResized(size_t size) const
...
+        __m128d value = _mm_load_sd(static_cast<const double *>(src));
+        _mm_store_sd(static_cast<double *>(dst), value);
+    }
+    else if constexpr (sizeof(T) == 16)
+    {
+        __m128i value = _mm_loadu_si128(static_cast<const __m128i *>(src));
+        _mm_storeu_si128(static_cast<__m128i *>(dst), value);
+    }
+#else
+    memcpy(dst, src, sizeof(T));
```
Comment (alexey-milovidov) [inline dbms/src/Columns/ColumnDecimal.cpp]: This will work in every case.

File: dbms/src/Columns/ColumnDecimal.cpp [hunk 3]
```diff
@@ -117,6 +121,36 @@ MutableColumnPtr ColumnDecimal<T>::cloneResized(size_t size) const
...
+template <typename T>
+void ColumnDecimal<T>::insertData(const char * pos, size_t /*length*/)
+{
+    data.push_back(T());
+
+    const void * src = pos;
+    void * dst = &data.back();
+
+#if __SSE2__
+    /// prevent aligned SSE load/store.
```
Comment (alexey-milovidov) [inline dbms/src/Columns/ColumnDecimal.cpp]: There are better solutions:
- replace non-portable `__int128_t` to `boost::int128_t`;
- figure out why `pos` is unaligned and fix it.

File: dbms/src/Columns/ColumnDecimal.cpp [hunk 4]
```diff
@@ -117,6 +121,36 @@ MutableColumnPtr ColumnDecimal<T>::cloneResized(size_t size) const
     return std::move(res);
 }
 
+template <typename T>
+void ColumnDecimal<T>::insertData(const char * pos, size_t /*length*/)
+{
+    data.push_back(T());
```
Comment (alexey-milovidov) [inline dbms/src/Columns/ColumnDecimal.cpp]: No need for this.

File: dbms/src/Columns/ColumnDecimal.cpp (hunk 3 above)
Comment (4ertus2) [reply dbms/src/Columns/ColumnDecimal.cpp]: pos is unaligned cause of Aggregator that doesn't care about aligment. It uses our UInt128 for any 128-bit types. It looks like there would be no profit if we replace __int128 and not replace UInt128 and UInt256.

File: dbms/src/Columns/ColumnDecimal.cpp (hunk 2 above)
Comment (4ertus2) [reply dbms/src/Columns/ColumnDecimal.cpp]: Yup. I have removes all SSE stuff. It's probably better to make tiny memcpy from library inlined function and use it.

**Pull request merged**